### PR TITLE
Implement byte[] as array of bytes

### LINKED ISSUE #169: Implement `byte[]` as array of bytes
This  implements `byte[]` using a specialized representation of arrays that represents the array as an array of bytes.

We have to deal with the fact that an `byte[] <: any[]`, so code dealing with `any[]` will have to deal with the (unlikely possibility) that it gets an array of bytes rather than a tagged pointer. (This is now handled using exactness.)

This depends on #168.

### PR DESCRIPTION
**Pull request by manuranga**
Fixes #169

### CODE REVIEW

File: runtime/list.c [hunk 1]
```diff
@@ -21,13 +21,20 @@ static inline bool listDescAllowsTaggedAt(ListDescPtr ldp, int64_t index, Tagged
     return memberTypeContainsTagged(mt, val);
 }
 
-ListPtr _bal_list_construct(ListDescPtr desc, int64_t capacity) {
+ListPtr _bal_list_generic_construct(ListDescPtr desc, int64_t capacity) {
```
Comment (jclark) [inline runtime/list.c]: `_generic_` isn't really the right word in this. You use it for int arrays also, right? It's for anything that has members of size 8.

There's a bit of a mismatch between `generic` as used in the ListDesc (which means any list) and `generic` in these `list_generic` functions, where it means tagged ptr.

File: runtime/list.c [hunk 2]
```diff
@@ -2,7 +2,7 @@
 
 #include <string.h>
 
-#define ARRAY_LENGTH_MAX ((int64_t)(INT64_MAX/sizeof(TaggedPtr)))
+#define GENERIC_ARRAY_LENGTH_MAX ((int64_t)(INT64_MAX/sizeof(TaggedPtr)))
```
Comment (jclark) [inline runtime/list.c]: Leave this as is. We are enforcing it for all arrays.

File: compiler/modules/nback/structure.bal [hunk 3]
```diff
@@ -314,14 +345,27 @@ function buildSpecializedListSet(llvm:Builder builder, Scaffold scaffold, llvm:V
...
-    builder.store(val, builder.getElementPtr(array, [llvm:constInt(LLVM_INT, 0), index], "inbounds"));
+    array = builder.bitCast(array, heapPointerType(llvm:arrayType(repr.memberHeapType, 0)));
+    llvm:Value heapVal = convertValue(builder, val, repr.memberType, repr.memberHeapType);
+    builder.store(heapVal, builder.getElementPtr(array, [llvm:constInt(LLVM_INT, 0), index], "inbounds"));
     builder.br(bbJoin);
     builder.positionAtEnd(bbTaggedSet);
     return bbJoin;
 }
 
+function convertValue(llvm:Builder builder, llvm:Value value, llvm:Type fromTy, llvm:Type toTy) returns llvm:Value {
```
Comment (jclark) [inline compiler/modules/nback/structure.bal:380]: I think it's better to have two functions listReprConvertTo/FromHeapType(ListRepr, Builder, Value)

**Review by jclark — CHANGES_REQUESTED** [review]

File: runtime/list.c (hunk 1 above)
Comment (manuranga) [reply runtime/list.c]: Ya, I was considering `_bal_list_construct_64` and `_bal_list_construct_8`.
`generic` in `GenericArray` means "not specific", and `generic` in `list_generic` means tagged.

File: runtime/list.c (hunk 2 above)
Comment (manuranga) [reply runtime/list.c]: I meant to increase the max size just for bytes up to INT64_MAX. Now I see I haven't done it correctly (should have changed `_bal_array_grow`), I'll revert this for now. 

But shouldn't we increase the max size for bytes?

File: runtime/list.c (hunk 1 above)
Comment (jclark) [reply runtime/list.c]: I think on the C side, it will make more sense to do this in terms of bytes (i.e. sizeof) rather than bits, so _8 and _1.

I think we should probably change list_generic to list_tagged.  But let's do that in a followup PR.

File: runtime/list.c (hunk 2 above)
Comment (jclark) [reply runtime/list.c]: I see no point in increasing it. Just useless extra complexity.

We cannot go up to INT64_MAX anyway since we are using top 8 bits in pointers as a tag.

File: runtime/list.c [hunk 4]
```diff
@@ -55,13 +64,23 @@ TaggedPtr _bal_list_int_array_get_tagged(TaggedPtr p, int64_t index) {
 }
 
 int64_t _bal_list_int_array_get_int(TaggedPtr p, int64_t index) {
-    // SUBSET should only be called when subset supports int subtype arrays
-    fprintf(stderr, "warning: access int array via int type desc func pointer\n");
```
Comment (jclark) [inline runtime/list.c]: Why is this needed for this PR?

File: compiler/modules/nback/structure.bal [hunk 5]
```diff
@@ -31,7 +31,16 @@ final RuntimeFunction mappingIndexedSetFunction = {
 };
 
 final RuntimeFunction listConstructFunction = {
```
Comment (jclark) [inline compiler/modules/nback/structure.bal]: This needs to be listConstruct8Function to be consistent with existing naming patterns

File: compiler/modules/nback/structure.bal [hunk 6]
```diff
@@ -31,7 +31,16 @@ final RuntimeFunction mappingIndexedSetFunction = {
...
+    name: "list_construct_8",
+    ty: {
+        returnType: heapPointerType(llListType),
+        paramTypes: [llvm:pointerType(llStructureDescType), LLVM_INT]
+    },
+    attrs: []
+};
+
+final RuntimeFunction byteListConstructFunction = {
+    name: "list_construct_1",
```
Comment (jclark) [inline compiler/modules/nback/structure.bal:43]: listConstruct1Function or change the C function name...

**Review by jclark — CHANGES_REQUESTED** [review]
Looks good.

File: runtime/list.c [hunk 7]
```diff
@@ -189,8 +210,7 @@ PanicCode _bal_list_generic_set_int(TaggedPtr p, int64_t index, int64_t val) {
 }
 
 PanicCode _bal_list_int_array_set_tagged(TaggedPtr p, int64_t index, TaggedPtr val) {
-    ListPtr lp = taggedToPtr(p);
-    if (!listDescAllowsTaggedAt(lp->desc, index, val)) {
+    if ((getTag(val) & UT_MASK) != TAG_INT) {
```
Comment (jclark) [inline runtime/list.c:213]: Shouldn't we have aa similar change to float_array_set_tagged?
